### DISCUSSION TITLE
3D Anaglyph Support

### DISCUSSION THREAD

**Discussion by yashchitroda** [iina/iina#5223] [Ideas]
@low-batt : Can we have a anaglyph support natively. This would definitely make it all rounder go to player for macOS having all capability

Following Media Players does support it natively

1. Kodi
2. VLC

**Comment by low-batt** [comment]
I'm a junior IINA developer. This would be something the senior developers would have to want to implement. But here are some thoughts from me…

For audio/video playback and subtitle rendering support IINA is using a library from the [mpv](https://github.com/mpv-player/mpv) project. The normal process is for mpv to add a feature. IINA will then add UI to control the new feature. The mpv issue https://github.com/mpv-player/mpv/issues/2583 requests some 3D support. There is also the closed issues https://github.com/mpv-player/mpv/issues/2413 and https://github.com/mpv-player/mpv/issues/1045.

On the IINA side there is issue #4272. You can see some replies from me on how it might be possible to get this to work in IINA using FFmpeg filters as indicated in the mpv issues.

**Comment by yashchitroda** [comment]
@low-batt I could see adding video filters turns of the hardware decoding, is this expected, if so can we have better alternative now of in future moving forward to add a native anaglyph support in IINA? what could be the work around now in IINA.

**Comment by low-batt** [comment]
Try the following:
- Start IINA
- Click on  `Settings…` under the `IINA` menu
- The settings panel appears
- On the left side of the panel click on `Video/Audio`
- In the `Video` section change the `Hardware decoder` setting to be `Auto (Copy)`

This IINA setting is merely changing the setting of the mpv option [hwdec](https://mpv.io/manual/stable/#options-hwdec). From the mpv manual:
> **--hwdec=<api1,api2,...|no|auto|auto-safe|auto-copy>**
> …
> **auto-copy:**	enable best hw decoder with copy-back (see below)
> …
> `auto-copy` selects only modes that copy the video data back to system memory after decoding. This selects modes like `vaapi-copy` (and so on). If none of these work, hardware decoding is disabled. This mode is usually guaranteed to incur no additional quality loss compared to software decoding (assuming modern codecs and an error free video stream), and will allow CPU processing with video filters. This mode works with all video filters and VOs.
>
> Because these copy the decoded video back to system RAM, they're often less efficient than the direct modes, and may not help too much over software decoding if you are short on CPU resources.

The important part of the manual entry is _will allow CPU processing with video filters_. Hardware acceleration is oriented towards normal playback. The most optimal version of it does not support filtering. The `Auto (Copy)` setting will still use hardware and should allow filtering.

If you read the entire manual entry you will find disappointing things about hardware acceleration. It has a history of being so defective that unlike IINA, mpv still defaults to always using software decoding. See the "Quality reduction with hardware decoding" section of the manual entry for some of the problems.

As an example, currently IINA is working around [FFmpeg](https://ffmpeg.org/) ticket [9599](https://trac.ffmpeg.org/ticket/9599) "VideoToolbox VP9 hwaccel freezes ffmpeg" by removing `vp9` from the mpv option [hwdec-codecs](https://mpv.io/manual/stable/#options-hwdec-codecs) when running on Macs that have Intel chips. This workaround addresses IINA issue #4486 while we are waiting for FFmpeg to fix the problem. So at this time hardware decoding of videos encoded using [VP9](https://en.wikipedia.org/wiki/VP9) is disabled in IINA on Intel Macs.

**Reply by yashchitroda** [reply]
https://github.com/iina/iina/discussions/5223#discussioncomment-11047960

**Comment by yashchitroda** [comment]
@low-batt what command should I add in vedio filter to achieve this sbs to anaglyph?
<img width="472" alt="Screenshot 2024-10-25 at 10 10 38 AM" src="https://github.com/user-attachments/assets/2d467d61-cdd2-463a-baf5-8c2e955c8308">

**Comment by low-batt** [comment]
Sorry I'm slow to respond. Very overloaded. This is a FFmpeg filter, not a mpv filter, so something like this:

![filter](https://github.com/user-attachments/assets/585b86d8-9e14-41d0-84fa-e1d93de4d55a)

As I don't know anything about 3D I'm not sure of which in/out values are appropriate. The choices are documented [here](https://ffmpeg.org/ffmpeg-filters.html#stereo3d).  The FFmpeg [Stereoscopic](https://trac.ffmpeg.org/wiki/Stereoscopic) page shows examples of what that filter can do.

After you add the filter you can save it with a name for later reuse.

**Comment by yashchitroda** [comment]
@low-batt can you suggest fields names here for mpv?

**Reply by low-batt** [reply]
I do not understand the question. What field names? Fields for a mpv filter? What mpv filter? Are you referring to the mpv filter for properly showing subtitles?

**Reply by yashchitroda** [reply]
@low-batt not for subtitles, but to achieve sbs to anaglyph what would be the Filter name, value to achieve this either for mpv, ffmpeg in IINA vf section

**Reply by low-batt** [reply]
That is what the screenshot above shows. Select `Custom (FFmpeg)`. Then fill in `stereo3d` for `Filter name` and `in=sbs2l:out=arbg` for the `Filter value`. That tells mpv to add the FFmpeg filter `stereo3d` with those parameters. Those parameters tell the filter the input is `side by side parallel (left eye left, right eye right)` and the output should be `anaglyph red/blue gray (red filter on left eye, blue filter on right eye)`.

**Reply by yashchitroda** [reply]
@low-batt what would be the Filter Name, Filter Value for custom FFmpeg, or mpv if I need to watch **H-SBS in normal 2D**

**Comment by yashchitroda** [comment]
This worked for me. also able to toggle using shortcut key Thanks @low-batt !! 
<img width="484" alt="Screenshot 2024-10-28 at 1 43 05 PM" src="https://github.com/user-attachments/assets/e650870c-0c14-4746-8607-94c27617ad1f">

**Comment by low-batt** [comment]
Great to hear you got it working!
